feat: add SD-JWT VC Type Metadata (#220)
oauth-wg/oauth-sd-jwt-vc: draft-ietf-oauth-sd-jwt-vc.md

## JWT VC Issuer Metadata Validation

The `issuer` value returned MUST be identical to the `iss` value of the
JWT. If these values are not identical, the data contained in the response
MUST NOT be used.

# Type Metadata {#type-metadata}

> A type is associated with metadata defining, for example, information about the type, a schema defining which claims MAY or MUST appear in the SD-JWT VC, and how they can be displayed.

Reviewer: This is qualified with "for example" but the type metadata in the PR currently doesn't define this stuff. I'm still unsure about leaving portions of sd-jwt-vc-types like claims and display out of this PR. But should we at least track that in an issue?

Reviewer: created #224

> This specification defines Type Metadata that can be associated with a type of a SD-JWT VC as well as a method for retrieving the Type Metadata and processing rules. This Type Metadata is intended to be used, among other things, for the following purposes:

Reviewer: ```suggestion
This section defines Type Metadata that can be associated with a type of a SD-JWT VC as well as a method for retrieving the Type Metadata and processing rules. This Type Metadata is intended to be used, among other things, for the following purposes:
```

 * Developers of Issuers and Verifiers can use the Typ Metadata to understand the
   semantics of the type and the associated rules. While in some cases,
   Issuers are the parties that define types (credential formats), this is
   not always the case. For example, a type can be defined by a
   standardization body or a community.
 * Verifiers can use the metadata to determine whether a credential is valid
   according to the rules of the type. For example, a Verifier can check

>    whether a credential contains all required claims and whether the claims

Reviewer: similar comment as https://github.com/oauth-wg/oauth-sd-jwt-vc/pull/220/files#r1548311395 here

   are selectively disclosable.
 * Wallets can use the metadata to display the credential in a way that is
   consistent with the Issuer's intent.

> Applications using this specification are called "Consumers" in the following. This typically includes Issuers, Verifiers, and Wallets.

Reviewer: ```suggestion
Applications using Type Metadata defined in this specification are called "Consumers" in the following. This typically includes Issuers, Verifiers, and Wallets.
```

  "description":"This is our development version of the education credential. Don't panic.",
  "extends":"https://galaxy.example.com/galactic-education-credential-0.9",
  "extends#integrity":"sha256-9cLlJNXN-TsMk-PmKjZ5t0WRL5ca_xGgX3c1VLmXfh-WRL5"
}
```

## Type Metadata Format {#type-metadata-format}

The Type Metadata document MUST be a JSON object. The following properties are
defined:

* `name`
  * OPTIONAL. A human-readable name for the type, intended for developers reading
  the JSON document.
* `description`
  * OPTIONAL. A human-readable description for the type, intended for
  developers reading the JSON document.
* `extends`
  * OPTIONAL. A URI of another type that this type extends, as described in
  (#extending-type-metadata).

## Extending Type Metadata {#extending-type-metadata}

A type can extend another type. The extended type is identified by the URI in
the `extends` property. Consumers MUST retrieve and process
Type Metadata for the extended type before processing the Type Metadata for the extending
type.

The extended type MAY itself extend another type. This can be used to create a
chain or hierarchy of types. The security considerations described in

> (#circular-extends) apply in order to avoid problems with circular dependencies.

Reviewer 2: I do believe we need a bit more text here on how exactly extending works:
- Can we only introduce new claims and re-use the parts we are extending (which is what I would assume so far?)
- Can we also override existing information?

Reviewer 3: Indeed, but that should be defined in the context (and the PR) of the particular property, e.g., the display property or the schema.

base64url-encoded Type Metadata documents as defined in this specification.
Multiple documents MAY be included for providing a whole chain of types to the
Consumer (see (#extending-type-metadata)).

A Consumer of a credential MAY use the documents in the `vctm`
array instead of retrieving the respective Type Metadata elsewhere as follows:

 * When resolving a `vct` in a credential, the Consumer MUST ensure
   that the `vct` claim in the credential matches the one in the Type Metadata
   document, and it MUST verify the integrity of the Type Metadata document as
   defined in (#document-integrity). The Consumer MUST NOT use the Type Metadata if no hash for integrity protection was provided in `vct#integrity`.
 * When resolving an `extends` property in a Type Metadata document, the Consumer MUST ensure that the value of the `extends` property in the
   Type Metadata document matches that of the `vct` in the Type Metadata document, and it MUST verify the integrity of the Type Metadata document as defined in
   (#document-integrity). The Consumer MUST NOT use the Type Metadata if no hash for integrity protection was provided.

> # Document Integrity {#document-integrity}

Reviewer: This should maybe be a subsection of Type Metadata so... ?
```suggestion
## Document Integrity {#document-integrity}
```

Reviewer 3: It is meant as a mechanism that can be used for multiple claims defined in the spec. If we only need it for the Type Metadata claims, this should be a subsection. Otherwise, a top-level section.

Author: It should be a top-level section imo since it is a general mechanism.

Author: agreed on the april 2nd call that it should stay a top-level claim.

## JWT VC Issuer Metadata Validation

The `issuer` value returned MUST be identical to the `iss` value of the
JWT. If these values are not identical, the data contained in the response
MUST NOT be used.

# Type Metadata {#type-metadata}

> A type is associated with rules defining which claims may or must appear in the SD-JWT VC and whether they may, must, or must not be selectively disclosable.

Reviewer 3: ```suggestion
A type is associated with metadata defining, for example, information about the type, a schema defining which claims may or must appear in the SD-JWT VC, and how they can be displayed.
```

Author: ```suggestion
A type is associated with metadata defining, for example, information about the type, a schema defining which claims MAY or MUST appear in the SD-JWT VC, and how they can be displayed.
```

This specification defines Type Metadata that can be associated with a type of a SD-JWT VC as well as a method for retrieving the Type Metadata and processing rules. This Type Metadata is intended to be used, among other things, for the following purposes:

 * Developers of Issuers and Verifiers can use the Typ Metadata to understand the
   semantics of the type and the associated rules. While in some cases,
   Issuers are the parties that define types (credential formats), this is
   not always the case. For example, a type can be defined by a
   standardization body or a community.
 * Verifiers can use the metadata to determine whether a credential is valid
   according to the rules of the type. For example, a Verifier can check
   whether a credential contains all required claims and whether the claims
   are selectively disclosable.
 * Wallets can use the metadata to display the credential in a way that is
   consistent with the Issuer's intent.

> Applications using this specification are called "consuming applications" in the following. This typically includes Issuers, Verifiers, and Wallets.

Reviewer 3: We might also just call them "consumers".

Author: I like that better.

A type can extend another type. The extended type is identified by the URI in
the `extends` property. The consuming application MUST retrieve and process
Type Metadata for the extended type before processing the Type Metadata for the extending
type.

The extended type MAY itself extend another type. This can be used to create a
chain or hierarchy of types. The security considerations described in
(#circular-extends) apply in order to avoid problems with circular dependencies.

## Retrieving Type Metadata {#retrieving-type-metadata}

### From a URL in the `vct` Claim {#retrieval-from-vct-claim}

A URI in the `vct` claim can be used to express a type. If the
type is a URL using the HTTPS scheme, Type Metadata can be retrieved from the URL
`https://<authority>/.well-known/vct/<type>`, i.e., by inserting
`/.well-known/vct` after the authority part of the URL.

The Type Metadata is retrieved using the HTTP GET method. The response MUST be a JSON
object as defined in (#type-metadata-format).

If the claim `vct#integrity` is present in the SD-JWT VC, its value
`vct#integrity` MUST be an "integrity metadata" string as defined in Section (#document-integrity).

> ### From a Registry {#retieval-from-registry}

Reviewer 3: ```suggestion
### From a Registry {#retrieval-from-registry}
```
